src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        // Take advantage of Python's surrogateescape encoding to support arbitrary byte
+        // sequences instead of limiting ourselves to UTF-8
+
+        // All targets besides windows support the std::os::unix::ffi::OsStrExt API:
+        // https://doc.rust-lang.org/src/std/sys_common/mod.rs.html#59
+        #[cfg(not(windows))]
+        {
+            let bytes = std::os::unix::ffi::OsStrExt::as_bytes(self);
+
+            let ptr = bytes.as_ptr() as *const c_char;
+            let len = bytes.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is this the appropriate function? Does it use the surrogateescape error
+                // handler? Should it be PyUnicode_DecodeLocaleAndSize instead maybe?
+                let pystring = ffi::PyUnicode_DecodeFSDefaultAndSize(ptr, len);
+                py.from_owned_ptr::<PyString>(pystring).into()
+            }
+        }
+
+        #[cfg(windows)]
+        {
+            let wstr: Vec<u16> = std::os::windows::ffi::OsStrExt::encode_wide(self).collect();
+
+            let ptr = wstr.as_ptr();
+            let len = wstr.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is there some better thing to do here than panic? Some last resort thing to
+                // try?
+                py.from_owned_ptr_or_err::<PyString>(ffi::PyUnicode_FromWideChar(ptr, len))
+                    .expect("OsStr conversion to Python string via FromWideChar failed")
+                    .into()
+            }
+        }
+    }
+}
+
+// It's not possible to implement FromPyObject for OsStr on Windows, because Python's wchar
+// representation can't be directly mapped to OsStr's one
+// HELP: should this implementation be here at all? It may be confusing for PyO3 users when a trait
+// is implemented only on some platforms
+#[cfg(not(windows))]
+impl<'a> FromPyObject<'a> for &'a OsStr {
+    fn extract(ob: &'a PyAny) -> PyResult<Self> {
+        let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+        let fs_encoded_bytes: &crate::types::PyBytes = unsafe {
+            ob.py()
+                .from_borrowed_ptr(ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
+        };
+
+        Ok(std::os::unix::ffi::OsStrExt::from_bytes(
+            fs_encoded_bytes.as_bytes(),
+        ))
+    }
+}
+
+impl FromPyObject<'_> for OsString {
+    fn extract(ob: &PyAny) -> PyResult<Self> {
+        #[cfg(not(windows))]
+        {
+            let os_str: &OsStr = ob.extract()?;
+            Ok(os_str.to_os_string())
+        }
+
+        #[cfg(windows)]
+        {
+            let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+            // Take the quick and easy shortcut if UTF-8
+            if let Ok(utf8_string) = pystring.to_str() {
+                return Ok(utf8_string.to_owned().into());
+            }
+
+            // The Windows way involves two allocations+copies, unfortunately.
```

Comment:
Technically the unix one does too (just that one of them is `PyBytes` which makes it less obvious). 😄